### PR TITLE
Add functionality to reconstruct splitted identifier

### PR DESCRIPTION
**Pull request by EgorBu**
* metadata for identifier reconstruction
* update logic for splitting
* logic for reconstruction of identifier
* test
Signed-off-by: egor <egor@sourced.tech>

### CODE REVIEW

File: sourced/ml/algorithms/token_parser.py [hunk 1]
```diff
@@ -1,29 +1,50 @@
+from enum import Enum
 import re
 
 import Stemmer
 
 
+class TokenMeta(Enum):
```
Comment (zurk) [inline sourced/ml/algorithms/token_parser.py]: More looks like `TokenStyle` if we call part of the identifier as Token. Or even `SubtokenStyle` to not confuse with Identifier as a code token.

File: sourced/ml/algorithms/token_parser.py [hunk 2]
```diff
@@ -1,29 +1,50 @@
...
+    """Metadata that should allow to reconstruct initial identifier from a list of tokens."""
+    DELIMITER = 1
+    TOKEN_UPPER = 2
+    TOKEN_LOWER = 3
+    TOKEN_STARTS_WITH_UPPER = 4
```
Comment (zurk) [inline sourced/ml/algorithms/token_parser.py]:
```suggestion
    TOKEN_CAPITALIZE = 4
```
It corresponds to `"abc".capitalize()`

File: sourced/ml/algorithms/token_parser.py [hunk 3]
```diff
@@ -1,29 +1,50 @@
...
+    #                                                  'very', '_', 'strange']
+    # NAME_BREAKUP_RE.split(token) -> ['Var', 'WithStrangeNAMING', 'very', 'strange']
+    STEM_THRESHOLD = 6  # We do not stem split parts shorter than or equal to this size.
+    MAX_TOKEN_LENGTH = 256  # We cut identifiers longer than this value.
+    MIN_SPLIT_LENGTH = 3  # We do not split source code identifiers shorter than this value.
+    DEFAULT_SINGLE_SHOT = False  # True if we do not want to join small identifiers to next one.
     # Example: 'sourced.ml.algorithms' -> ["sourc", "sourcedml", "algorithm", "mlalgorithm"].
     # if True we have only ["sourc", "algorithm"].
     # if you do not want to filter small tokens set min_split_length=1.
+    METADATA = False  # whether yield metadata that can be used to reconstruct initial identifier
```
Comment (zurk) [inline sourced/ml/algorithms/token_parser.py]: This one can be confused with model metadata. So better to name `SAVE_TOKEN_STYLE`

File: sourced/ml/algorithms/token_parser.py [hunk 4]
```diff
@@ -76,6 +97,22 @@ def stem(self, word):
     def split(self, token):
         token = token.strip()[:self.max_token_length]
 
+        def meta_decorator(func):
+            if self._metadata:
+                def decorated_func(name):
```
Comment (zurk) [inline sourced/ml/algorithms/token_parser.py]:
```suggestion
                @functools.wraps(func)
                def decorated_func(name):
```

File: sourced/ml/algorithms/token_parser.py [hunk 5]
```diff
@@ -76,6 +97,22 @@ def stem(self, word):
...
+                    if name.isupper():
+                        meta = TokenMeta.TOKEN_UPPER
+                    elif name.islower():
+                        meta = TokenMeta.TOKEN_LOWER
+                    else:
+                        meta = TokenMeta.TOKEN_STARTS_WITH_UPPER
```
Comment (zurk) [inline sourced/ml/algorithms/token_parser.py]:
```suggestion
                        assert name[0].isupper() and name[1:].islower()
                        meta = TokenMeta.TOKEN_STARTS_WITH_UPPER
```

File: sourced/ml/algorithms/token_parser.py [hunk 6]
```diff
@@ -1,29 +1,50 @@
...
     Common utilities for splitting and stemming tokens.
     """
-    NAME_BREAKUP_RE = re.compile(r"[^a-zA-Z]+")  #: Regexp to split source code identifiers.
-    STEM_THRESHOLD = 6  #: We do not stem split parts shorter than or equal to this size.
-    MAX_TOKEN_LENGTH = 256  #: We cut identifiers longer than this value.
-    MIN_SPLIT_LENGTH = 3  #: We do not split source code identifiers shorter than this value.
-    DEFAULT_SINGLE_SHOT = False  #: True if we do not want to join small identifiers to next one.
+    # Regexp to split source code identifiers
+    NAME_BREAKUP_RE = re.compile(r"[^a-zA-Z]+")
+    NAME_BREAKUP_KEEP_DELIMITERS_RE = re.compile(r"([^a-zA-Z]+)")  # ... and keep delimiters
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/token_parser.py]: This is not required at all. Everything which is between groups detected by `NAME_BREAKUP_RE` should be taken.

**Comment by vmarkovtsev** [comment]
My ideal implementation is **exactly** as I stated in https://github.com/src-d/style-analyzer/issues/662#issuecomment-470063925 Nothing is touched in the outside world, and I do believe that it is possible to implement this way.

File: sourced/ml/algorithms/token_parser.py (hunk 1 above)
Comment (vmarkovtsev) [reply sourced/ml/algorithms/token_parser.py]: \+1 for TokenStyle

File: sourced/ml/algorithms/token_parser.py [hunk 7]
```diff
@@ -1,29 +1,50 @@
...
+    DEFAULT_SINGLE_SHOT = False  # True if we do not want to join small identifiers to next one.
     # Example: 'sourced.ml.algorithms' -> ["sourc", "sourcedml", "algorithm", "mlalgorithm"].
     # if True we have only ["sourc", "algorithm"].
     # if you do not want to filter small tokens set min_split_length=1.
+    METADATA = False  # whether yield metadata that can be used to reconstruct initial identifier
 
     def __init__(self, stem_threshold=STEM_THRESHOLD, max_token_length=MAX_TOKEN_LENGTH,
-                 min_split_length=MIN_SPLIT_LENGTH, single_shot=DEFAULT_SINGLE_SHOT):
+                 min_split_length=MIN_SPLIT_LENGTH, single_shot=DEFAULT_SINGLE_SHOT,
+                 metadata=METADATA):
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/token_parser.py]: `metadata` -> `include_delimiters`

File: sourced/ml/algorithms/token_parser.py [hunk 8]
```diff
@@ -1,29 +1,50 @@
...
         self._stemmer = Stemmer.Stemmer("english")
         self._stemmer.maxCacheSize = 0
         self._stem_threshold = stem_threshold
         self._max_token_length = max_token_length
         self._min_split_length = min_split_length
         self._single_shot = single_shot
+        self._metadata = metadata
+        if self._metadata:
+            assert self._single_shot
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/token_parser.py]: This is a ValueError

**Pull request merged**